core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/clustertemplate/ClusterTemplateV4RequestToClusterTemplateConverter.java

```diff
@@ -63,6 +67,16 @@ public ClusterTemplate convert(ClusterTemplateV4Request source) {
         if (StringUtils.isEmpty(source.getDistroXTemplate().getEnvironmentName())) {
             throw new BadRequestException("The environmentName cannot be null.");
         }
+        Set<InstanceGroupV1Request> gateways = source.getDistroXTemplate().getInstanceGroups()
+                .stream()
+                .filter(ig -> ig.getType().equals(InstanceGroupType.GATEWAY))
```

Comment:
Potential NPE. I would either swap the `equals()` (so `InstanceGroupType.GATEWAY.equals(ig.getType())`) or filter out `ig.getType() == null` values.

Option 3: Add `@NotNull` to `InstanceGroupV1Base.type` (https://github.com/hortonworks/cloudbreak/blob/5a50ac58b3f8ed657d6f2934bcbcac98389ed4c6/core-api/src/main/java/com/sequenceiq/distrox/api/v1/distrox/model/instancegroup/InstanceGroupV1Base.java#L32-L33). I do not know why it is nullable in the first place. :smile: